httprequest_lego_provider/dns.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""DNS utiilities."""
+
+import io
+from pathlib import Path
+from tempfile import TemporaryDirectory
+
+from git import Git, GitCommandError, Repo
+
+from .settings import DNS_REPOSITORY_URL, SSH_IDENTITY_FILE
+
+FILENAME_SUFFIX = ".domain"
+SPLIT_DNS_REPOSITORY_URL = DNS_REPOSITORY_URL.rsplit("@", 1)
+REPOSITORY_BASE_URL = SPLIT_DNS_REPOSITORY_URL[0]
+REPOSITORY_BRANCH = SPLIT_DNS_REPOSITORY_URL[1] if len(SPLIT_DNS_REPOSITORY_URL) > 1 else None
+RECORD_CONTENT = "{record} 600 IN TXT \042{value}\042\n"
+SSH_EXECUTABLE = f"ssh -i {SSH_IDENTITY_FILE}"
+
+
+class DnsSourceUpdateError(Exception):
+    """Exception for DNS update errors."""
+
+
+def write_dns_record(dns_record: str, value: str) -> None:
+    """Write a DNS record following the canonical-is-dns-configs specs if it doesn't exist.
+
+    Args:
+        dns_record: the DNS record to add.
+        value: ACME challenge for DNS record to add.
+
+    Raises:
+        DnsSourceUpdateError: if an error while updating the repository occurs.
+    """
+    with TemporaryDirectory() as tmp_dir, Git().custom_environment(GIT_SSH_COMMAND=SSH_EXECUTABLE):
+        try:
+            repo = Repo.clone_from(REPOSITORY_BASE_URL, tmp_dir, branch=REPOSITORY_BRANCH)
+            splitted_record = dns_record.rsplit(".", 2)
+            subdomain = splitted_record[0]
+            domain = ".".join(splitted_record[1:]) if splitted_record else "."
```

Comment:
We should proably add a comment here specifying that we expect the domain to ALWAYS be prefixed by "_acme_challenge", therefore the case of `dns_record = "canonical.com"` shouldn't happen.

In fact we should even validate that the DNS record starts with "_acme_challenge" to avoid potential issues

Reply:
Perhaps we should append it to the entry?

Reply:
We it's not the user providing it but we get it from the fqdn field sent by "lego" and I expect lego to always prepend it. But as usual never trust the input, so better check in case this API is called by something(/one) else than Lego

Reply:
Noted. I'll do this in a follow up PR if that's ok with you, as we might change the current API